File: first-last-element.py
# Time Complexity : O(log n)
# Space Complexity : O(1)
# Did this code successfully run on Leetcode : Yes
# Any problem you faced while coding this : No

from typing import List


class Solution:
    def searchRange(self, nums: List[int], target: int) -> List[int]:

        low = 0
        high = len(nums)-1
        first = last = -1

        # find first index
        while (low <= high):
            mid = (low+high)//2

            # if target is found
            if (nums[mid] == target):
                # check if its the lowest index
                if mid == 0 or nums[mid-1] < target:
                    first = mid
                    break
                # find lowest index
                else:
                    high = mid-1
            elif nums[mid] > target:
                high = mid-1
            else:
                low = mid+1

        low = first + 1
        last = first
        high = len(nums)-1

        # if target exists in array find last index
        if first != -1:
            while (low <= high):
                mid = (low+high)//2
                # when target found
                if (nums[mid] == target):
                    # check if its the max index
                    if mid == len(nums)-1 or nums[mid+1] > target:
                        last = mid
                        break
                    # find max index
                    else:
                        low = mid+1
                elif nums[mid] > target:
                    high = mid-1
                else:
                    low = mid+1

        return [first, last]

File: first_last_element.java
// Time Complexity :O(log n)
// Space Complexity : O(1) 
// Did this code successfully run on Leetcode : Yes
// Any problem you faced while coding this :


class Solution {

public int[] searchRange(int[] nums, int target) { 
  if( nums == null || nums.length == 0) {
       return new int[]{-1, -1}; 
     }

  int l = 0;
  int n = nums.length;
  int r = n -1;
 
  if (nums[0] > target || nums[n-1] < target){
     return new int[]{-1,-1}; 
  }

  int first  = firstBinarysearch(nums, l,r,target);
  if (first == -1)
  {
        return new int[]{-1,-1}; 
 
  }
  int last  = lastBinarysearch(nums,first,r,target);

  return new int[]{first,last};

    }


public int firstBinarysearch(int[] nums,int l ,int r,int target){
while(l<=r){
int mid = l + (r-l)/2;
if (nums[mid] == target){
    if( mid == 0 || nums[mid] > nums[mid-1])  
    return mid;
    else 
    r = mid-1;
}
 else if (nums[mid]>target){
    r = mid -1;
 }
 else l = mid+1;
}
return -1;
}


public int lastBinarysearch(int[] nums,int l ,int r,int target){
while(l<=r){
int mid = l + (r-l)/2;
if (nums[mid] == target){
    if( mid == nums.length-1 || nums[mid] < nums[mid+1]) { 
    return mid;
    }
    else{
    l = mid + 1;
    }
}
 else if (nums[mid]>target){
    r = mid -1;

 }
 else {l = mid+1;}
}
return -1;
}


}
